Source/DafnyTests/DafnyTestCase.cs

```diff
@@ -0,0 +1,417 @@
+        var compileTargets = new[] {"cs", "java", "go", "js"};
+
+        var justVerify = new Dictionary<string, object>(DafnyArguments) {
+          [DAFNY_COMPILE_OPTION] = "0"
+        };
+        yield return new DafnyTestSpec(SourcePath, justVerify, OtherFiles, DafnyTestCase.Expectation.NO_OUTPUT);
+        
+        foreach (var compileTarget in compileTargets) {
+          var withLanguage = new Dictionary<string, object>(DafnyArguments) {
+              [DAFNY_NO_VERIFY_OPTION] = "yes", 
+              [DAFNY_COMPILE_OPTION] = "4",
+              [DAFNY_COMPILE_TARGET_OPTION] = compileTarget
+            };
+          var specForLanguage = new DafnyTestSpec(SourcePath, withLanguage, OtherFiles, Expected);
+          if (CompileTargetOverrides.TryGetValue(compileTarget, out var compileTargetOverride)) {
+            yield return specForLanguage.ApplyOverride(compileTargetOverride);
+          } else {
+            yield return specForLanguage;
+          }
+        }
+      } else {
+        yield return this;
+      }
+    }
+
+    private DafnyTestSpec ApplyOverride(DafnyTestSpec otherSpec) {
+      var mergedArguments = new Dictionary<string, object>(DafnyArguments);
+      foreach (KeyValuePair<string, object> pair in otherSpec.DafnyArguments) {
+        mergedArguments[pair.Key] = pair.Value;
+      }
+      return new DafnyTestSpec(otherSpec.SourcePath, mergedArguments, OtherFiles.Concat(otherSpec.OtherFiles).ToList(), otherSpec.Expected);
+    }
+    
+    private DafnyTestSpec ResolveExpected() {
+      if (Expected == null) {
+        return new DafnyTestSpec(SourcePath, DafnyArguments, OtherFiles, 
+                                 new DafnyTestCase.Expectation(0, Path.GetFileName(SourcePath) + ".expect", null));
+      } else {
+        return this;
+      }
+    }
+    
+    private DafnyTestCase ToTestCase() {
+      var arguments = new []{ Path.GetRelativePath(TEST_ROOT, SourcePath) }
+        .Concat(DafnyArguments.Select(ConfigPairToArgument))
+        .Concat(OtherFiles.Select(otherFile => "comp/" + otherFile));
+      return new DafnyTestCase(arguments, Expected.Adjust());
+    }
+    
+    private static string ConfigPairToArgument(KeyValuePair<string, object> pair) {
+      if (pair.Value.Equals("yes")) {
+        return String.Format("/{0}", pair.Key);
+      } else {
+        return String.Format("/{0}:{1}", pair.Key, pair.Value);
+      }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator() {
+      return GetEnumerator();
+    }
+    
+    private IEnumerable<DafnyTestSpec> ExpandArguments() {
+      return DafnyArguments.Select(ExpandValue)
+                           .CartesianProduct()
+                           .Select(e => e.ToDictionary(p => p.Key, p => p.Value))
+                           .Select(args => new DafnyTestSpec(SourcePath, args, OtherFiles, Expected));
+    }
+
+    public static IEnumerable<string> Expand(object obj) {
+      if (obj is ForEachArgumentList forEach) {
+        return forEach;
```

Comment:
np because it takes a second to realize that `forEach` represents the list and not an argument within the list. I know we use `foreach` in other contexts, but I think this would make it a bit clearer here, at least.

```suggestion
      if (obj is ForEachArgumentList forEachArgumentList) {
        return forEachArgumentList;
```